### PR TITLE
Add LayoutLMv3

### LINKED ISSUE #16914: LayoutLMv3: Pre-training for Document AI with Unified Text and Image Masking
### Model description

LayoutLMV3 is the successor of the LayoutLM models. the models are specialized in multimodal document analysis tasks and achieve SOTA results on them.

The current [code](https://github.com/microsoft/unilm/blob/master/layoutlmv3/layoutlmft/models/layoutlmv3/modeling_layoutlmv3.py) for the LayoutLMV3 is implemented in huggingface format.

Since the code is already more similar to the Huggingface format, I am not sure whether integrating this model into this repo is necessary, but there will be an ease of use if it is available in the library itself.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

Model Implementation: https://github.com/microsoft/unilm/tree/master/layoutlmv3
Model Weights: https://huggingface.co/microsoft/layoutlmv3-base
Authors: Yupan Huang,Tengchao Lv, Lei Cui, Yutong Lu ,Furu Wei

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR implements LayoutLMv3. LayoutLMv3 doesn't require a Detectron2 backbone anymore (yay!).

The PR also includes an example script that can be used to reproduce results of the paper. 

Fixes #16914

To do:

- [x] fix remaining tokenizer tests. These are very black-boxy to me. Pinging @SaulLu here.
- [x] add model to doc tests
- [x] remove `is_detection` logic
- [x] Make sure the slow tests involving `PyTesseract` pass
- [x] Merge `add_layoutlmv3_simplify` branch

### CODE REVIEW

**Comment by SaulLu** [comment]
Just for the purpose to keep track of the current status.

As discussed offline I think the next step to "solve" the tokenization tests is to figure out how `["hello", "world"]` is tokenized in the original code: is it [0, 42891, 8331, 2] (['<s>', 'Ġhello', 'Ġworld', '</s>']) or [0, 20760, 232, 2]  (['<s>', 'hello', "world", '</s>']) or something else ? :blush:

**Comment by NielsRogge** [comment]
As seen [here](https://github.com/microsoft/unilm/blob/925de7a9ea500e992ec5de02ea193a5eb9d5aa26/layoutlmv3/examples/run_funsd_cord.py#L313), text is tokenized using RobertaTokenizer, where one provides `is_split_into_words=True`. Hence, "hello", "world" is tokenized as follows:

```
from transformers import RobertaTokenizer

tokenizer = RobertaTokenizer.from_pretrained("microsoft/layoutlmv3-base")

text = ["hello", "world"]

encoding = tokenizer(text, is_split_into_words=True)
```
So this results in [0, 20760, 232, 2].

**Comment by SaulLu** [comment]
Thanks for the clarification! I've opened a PR on your branch (https://github.com/NielsRogge/transformers/pull/38) which proposes several changes including 1) changing the default behaviour so that by default a space prefix is added and including all the changes needed to make it work and 2) some small changes to resolve several of the tests that were failing.

I wonder if we shouldn't just remove the option to set `add_prefix_space` to False because the result will not be satisfactory for decoding and I'm not sure we want to do any fancy tricks to make it "work". (Or at least we should log a message to warn the user that the option is risky).

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -546,8 +554,7 @@
     _import_structure["models.glpn"].append("GLPNFeatureExtractor")
     _import_structure["models.imagegpt"].append("ImageGPTFeatureExtractor")
     _import_structure["models.layoutlmv2"].append("LayoutLMv2FeatureExtractor")
-    _import_structure["models.layoutlmv2"].append("LayoutLMv2Processor")
-    _import_structure["models.layoutxlm"].append("LayoutXLMProcessor")
```
Comment (NielsRogge) [inline src/transformers/__init__.py:550]: The processors shouldn't be under a `is_vision_available`, so I removed them for LayoutLMv2 and LayoutXLM.

File: src/transformers/__init__.py [hunk 2]
```diff
@@ -2913,8 +2938,8 @@
         from .models.dpt import DPTFeatureExtractor
         from .models.glpn import GLPNFeatureExtractor
         from .models.imagegpt import ImageGPTFeatureExtractor
-        from .models.layoutlmv2 import LayoutLMv2FeatureExtractor, LayoutLMv2Processor
-        from .models.layoutxlm import LayoutXLMProcessor
```
Comment (NielsRogge) [inline src/transformers/__init__.py:2917]: Same here.

File: docs/source/en/_toctree.yml [hunk 3]
```diff
@@ -234,6 +234,8 @@
       title: LayoutLM
     - local: model_doc/layoutlmv2
       title: LayoutLMV2
+    - local: model_doc/layoutlmv3
+      title: LayoutLMV3
```
Comment (sgugger) [inline docs/source/en/_toctree.yml:238]: As usual, we need to add the model to the README too :-)

File: examples/research_projects/layoutlmv3/data_collator.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            * :obj:`False` or :obj:`'do_not_pad'` (default): No padding (i.e., can output a batch with sequences of
+              different lengths).
+        max_length (:obj:`int`, `optional`):
+            Maximum length of the returned list and optionally padding length (see above).
+        pad_to_multiple_of (:obj:`int`, `optional`):
+            If set will pad the sequence to a multiple of the provided value.
+            This is especially useful to enable the use of Tensor Cores on NVIDIA hardware with compute capability >=
+            7.5 (Volta).
+        label_pad_token_id (:obj:`int`, `optional`, defaults to -100):
+            The id to use when padding the labels (-100 will be automatically ignore by PyTorch loss functions).
```
Comment (sgugger) [inline examples/research_projects/layoutlmv3/data_collator.py:35]: Either update to the new format of docstings (no :obj: and :class:) or you can remove it entirely if easier.

File: src/transformers/models/layoutlmv3/configuration_layoutlmv3.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
+# coding=utf-8
+# Copyright Microsoft Research and The HuggingFace Inc. team. All rights reserved.
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/configuration_layoutlmv3.py:2]:
```suggestion
# Copyright 2022 Microsoft Research and The HuggingFace Inc. team. All rights reserved.
```

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 6]
```diff
@@ -0,0 +1,1298 @@
...
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+# Copied from transformers.models.vit.modeling_vit.to_2tuple
+def to_2tuple(x):
+    if isinstance(x, collections.abc.Iterable):
+        return x
+    return (x, x)
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:128]: I think we said we didn't want to use this anymore and directly code the test when we use it?

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 7]
```diff
@@ -0,0 +1,1298 @@
...
+
+class LayoutLMv3PatchEmbeddings(nn.Module):
+    """LayoutLMv3 image (patch) embeddings. This class also automatically interpolates the position embeddings for varying
+    image sizes."""
+
+    def __init__(self, config):
+        super().__init__()
+
+        image_size = to_2tuple(config.input_size)
+        patch_size = to_2tuple(config.patch_size)
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:139]:
```suggestion
        image_size = config.input_size if isinstance(config.input_size, collections.abc.Iterable) else (config.input_size, config.input_size)
        patch_size = config.patch_size if isinstance(config.patch_size, collections.abc.Iterable) else (config.patch_size, config.patch_size)
```

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 8]
```diff
@@ -0,0 +1,1298 @@
...
+        self.proj = nn.Conv2d(config.num_channels, config.hidden_size, kernel_size=patch_size, stride=patch_size)
+
+    def forward(self, pixel_values, position_embedding=None):
+        embeddings = self.proj(pixel_values)
+
+        if position_embedding is not None:
+            # interpolate the position embedding to the corresponding size
+            position_embedding = position_embedding.view(1, self.patch_shape[0], self.patch_shape[1], -1).permute(
+                0, 3, 1, 2
+            )
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:150]:
```suggestion
            position_embedding = position_embedding.view(1, self.patch_shape[0], self.patch_shape[1], -1)
            position_embedding = position_embedding.permute(0, 3, 1, 2)
```

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 9]
```diff
@@ -0,0 +1,1298 @@
...
+        )
+
+        self.x_position_embeddings = nn.Embedding(config.max_2d_position_embeddings, config.coordinate_size)
+        self.y_position_embeddings = nn.Embedding(config.max_2d_position_embeddings, config.coordinate_size)
+        self.h_position_embeddings = nn.Embedding(config.max_2d_position_embeddings, config.shape_size)
+        self.w_position_embeddings = nn.Embedding(config.max_2d_position_embeddings, config.shape_size)
+
+    def calculate_spatial_position_embeddings(self, bbox):
+        try:
+            assert torch.all(0 <= bbox) and torch.all(bbox <= 1023)
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:187]: No new assert in a modeling file please

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 10]
```diff
@@ -0,0 +1,1298 @@
...
+            self.rel_2d_pos_onehot_size = config.rel_2d_pos_bins
+            self.rel_pos_x_bias = nn.Linear(self.rel_2d_pos_onehot_size, config.num_attention_heads, bias=False)
+            self.rel_pos_y_bias = nn.Linear(self.rel_2d_pos_onehot_size, config.num_attention_heads, bias=False)
+
+    def relative_position_bucket(self, relative_position, bidirectional=True, num_buckets=32, max_distance=128):
+        ret = 0
+        if bidirectional:
+            num_buckets //= 2
+            ret += (relative_position > 0).long() * num_buckets
+            n = torch.abs(relative_position)
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:508]: Would love a better name for `n`

File: src/transformers/models/layoutlmv3/test.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+from transformers import LayoutLMv3Tokenizer
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/test.py:1]: This files shouldn't be included in the PR :-)

File: src/transformers/models/layoutlmv3/tokenization_layoutlmv3_fast.py [hunk 12]
```diff
@@ -0,0 +1,854 @@
...
+        words = text if text_pair is None else text_pair
+        assert boxes is not None, "You must provide corresponding bounding boxes"
+        if is_batched:
+            assert len(words) == len(boxes), "You must provide words and boxes for an equal amount of examples"
+            for words_example, boxes_example in zip(words, boxes):
+                assert len(words_example) == len(
+                    boxes_example
+                ), "You must provide as many words as there are bounding boxes"
+        else:
+            assert len(words) == len(boxes), "You must provide as many words as there are bounding boxes"
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/tokenization_layoutlmv3_fast.py:322]: No new asserts in the code base please :-)

File: tests/models/layoutlmv3/test_feature_extraction_layoutlmv3.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
+# coding=utf-8
+# Copyright 2021 HuggingFace Inc.
```
Comment (sgugger) [inline tests/models/layoutlmv3/test_feature_extraction_layoutlmv3.py:2]:
```suggestion
# Copyright 2022 HuggingFace Inc.
```

File: tests/models/layoutlmv3/test_feature_extraction_layoutlmv3.py [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+        self.assertEqual(
+            encoding.pixel_values.shape,
+            (
+                1,
+                3,
+                224,
+                224,
+            ),
+        )
```
Comment (sgugger) [inline tests/models/layoutlmv3/test_feature_extraction_layoutlmv3.py:221]:
```suggestion
        self.assertEqual(encoding.pixel_values.shape, (1, 3, 224, 224))
```

**Review by sgugger — APPROVED** [review]
Nice work adding this new model!
There are a lot of commented-out code in the `test_tokenization` file. Not sure if it's too fix or cleanup, but it should be removed before merging the PR. LGTM otherwise!

**Comment by ducviet00** [comment]
Hi @NielsRogge 

As the issue #13554 and PR #17092,  when `input_ids` is longer than model's `max_length`, it would be split into multiple inputs, but `pixel_values` still has 1 image. Are you going to fix this right now, or next PR?

How to reproduce
```python
from transformers import AutoProcessor, AutoModelForTokenClassification
from datasets import load_dataset
from PIL import Image
processor = AutoProcessor.from_pretrained("microsoft/layoutlmv3-large")
processor.feature_extractor.apply_ocr = False
model = AutoModelForTokenClassification.from_pretrained("microsoft/layoutlmv3-large")

words = ['hello' for i in range(1000)]
boxes = [[0, 1, 2, 3] for i in range(1000)]
encoding = processor(
    image, 
    text=words, 
    boxes=boxes,
    truncation=True,
    padding='max_length',
    return_overflowing_tokens=True, 
    return_tensors="pt"
)

print(encoding['input_ids'].shape) # torch.Size([2, 512])
print(encoding['pixel_values'].shape) #torch.Size([1, 3, 224, 224])
overflow_to_sample_mapping = encoding.pop('overflow_to_sample_mapping')
model(**encoding) 
# ---> RuntimeError: Sizes of tensors must match except in dimension 1.
# Expected size 4 but got size 1 for tensor number 1 in the list.
```

**Review by LysandreJik — COMMENTED** [review]
That's very impressive! Have the tokenization and test files been copied from others? I see the `# Copied from` statements only in the modeling file, it would likely greatly help reviewing if they were also in the other files which have copied parts of the code

File: tests/models/layoutlmv3/test_tokenization_layoutlmv3.py [hunk 15]
```diff
@@ -0,0 +1,2345 @@
...
+                    truncation="only_first",
+                    return_overflowing_tokens=True,
+                    # add_prefix_space=False,
+                )
+                # Overflowing tokens are handled quite differently in slow and fast tokenizers
+                if isinstance(tokenizer, LayoutLMv3TokenizerFast):
+                    truncated_sequence = information_first_truncated["input_ids"][0]
+                    overflowing_tokens = information_first_truncated["input_ids"][1]
+                    bbox = information_first_truncated["bbox"][0]
+                    overflowing_bbox = information_first_truncated["bbox"][0]
```
Comment (SaulLu) [inline tests/models/layoutlmv3/test_tokenization_layoutlmv3.py:2026]: A small note to keep in mind the ongoing discussion we had offline, I'm not sure I understand why the element in position 0 is taken and not the one in position 1. :slightly_smiling_face:

**Comment by sina-ehsani** [comment]
Thank you so much for your fantastic work. I was wondering if you plan to include the object detection task in LayoutLMv3 as well. I noticed that the [PubLayNet fine-tuned model weights](https://huggingface.co/HYPJUDY/layoutlmv3-base-finetuned-publaynet) have already been uploaded to HuggingFace, but I couldn't find any documentation on this capability in this repository.

**Review by LysandreJik — APPROVED** [review]
Impressive effort! LGTM!

And thanks for adding the `# Copied from` statements, makes the review easier.

**Comment by dcyoung** [comment]
@NielsRogge Thanks for this contribution! 
While testing the processor, I'm seeing extra padding on the resultant labels that I did not expect, have not experienced with older versions of layoutlmv2processor. 

```
import numpy as np
from transformers.models.auto.processing_auto import AutoProcessor

processor = AutoProcessor.from_pretrained(
    pretrained_model_name_or_path="microsoft/layoutlmv3-base",
    use_fast=True,
    add_prefix_space=True,
    apply_ocr=False,
)

# not batched
words = ["hello", "world"]
boxes = [[1, 2, 3, 4], [5, 6, 7, 8]]
word_labels = [1, 2]
image = np.zeros((224, 224, 3), dtype=np.uint8)
results = processor(
    image, words, boxes=boxes, word_labels=word_labels, return_tensors="pt"
)
for k, v in results.items():
    print(k, v.size())

labels = results.labels.squeeze().tolist()
print(labels)
```
output:
```
input_ids torch.Size([1, 8])
attention_mask torch.Size([1, 8])
bbox torch.Size([1, 8, 4])
labels torch.Size([1, 205])
pixel_values torch.Size([1, 3, 224, 224])
[-100, 1, -100, -100, -100, -100, 2, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100]
```

This happens beyond maximum seq length as well... where the labels will have a dimension seq_length + ~197
Is this expected?

**Comment by NielsRogge** [comment]
Hi @dcyoung,

thanks for taking a look. Actually you make a great point; I implemented it as the original implementation (where the authors label all visual tokens with -100 and just add a classifier on top of the entire `sequence_output`), however it makes a lot of sense to just simplify the code in `LayoutLMv3ForTokenClassification` and not make the processor do this. 

Thanks a lot!

**Comment by NielsRogge** [comment]
And hi @sina-ehsani, 

unfortunately I'm (for now) not planning to add the object detection part, because the framework being used (Mask R-CNN) is a ridiculous amount of code and it's not straightforward - for now - to add this to the Transformers library (as there's a "one model, one file" philosophy). So I'd advise to use the original repository for that. 

It may be that in the future we add this framework, but I'm actually much more a fan of simpler frameworks like DETR and YOLOS. It would be great if someone trained a YOLOS model initialized with the weights of the [Document Image Transformer (DiT)](https://huggingface.co/docs/transformers/model_doc/dit). I feel like you would get the same performance.

**Comment by divinit7** [comment]
Thank you so much for adding the model, I had a question on segment position embeddings. How do you create segment position embeddings during inference when the labels are unknown and are just bounding boxes from an ocr. In this [notebook](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/LayoutLMv3/Fine_tune_LayoutLMv3_on_FUNSD_(HuggingFace_Trainer).ipynb) the test set also contains segment level bounding box. I have trained a model on segment level embeddings on my use case and it doesn't perform well on token level 2D embeddings during inference.

**Comment by jordanparker6** [comment]
> YOLOS

Thanks for the idea. I will have a go at this. 

However, my understanding is that YOLOS / DETR are image only and do not include the token embeddings from text. Whereas LayoutLM leverages both the text and visual context. The content of a paragraph might be informative for a classification of an object boundary.

**Comment by jordanparker6** [comment]
@NielsRogge 

> 

This thread has lead me to hacking a model that combines the YolosLoss and YolosObjectDetection head with the LayoutLMv3Model to build a LayoutLMv3ObjectDetection prediction head.

Changes to the LayoutLMv3Config and LayoutLMv3FeatureExtractor had to be made to allow for this.

This approach avoids the Mask R-CNN discussed.

Is this something you would be interested in reviewing and integrating if I open a PR? 

Or does it deviate too significantly from the research paper?
